Support schema evolution across versions

naegelejd opened the pull request:
Introduces support for schema evolution across versions, including automated schema change detection and backward-compatible serialization (currently only C++ binary).

johnstairs commented on models/evolution/model_v0/_package.yml:
```diff
@@ -0,0 +1,8 @@
+namespace: EvoTest
+
+cpp:
+  sourcesOutputDir: ../../../cpp/evolution/v0/lib
```
In our other models, we have generated to a directory named `generated`. I think  we should be consistent.

johnstairs commented on cpp/test/generated/binary/protocols.cc:
```diff
@@ -706,7 +706,7 @@ template<typename T1, yardl::binary::Reader<T1> ReadT1, typename T2, yardl::bina
 namespace {
 template<typename K, yardl::binary::Writer<K> WriteK, typename V, yardl::binary::Writer<V> WriteV>
-[[maybe_unused]] void WriteAliasedMap(yardl::binary::CodedOutputStream& stream, basic_types::AliasedMap<K, V> const& value) {
+[[maybe_unused]] static void WriteAliasedMap(yardl::binary::CodedOutputStream& stream, basic_types::AliasedMap<K, V> const& value) {
```
Is the addition of `static` to make the internal linkage explicit (even though the function is in an unnamed namespace) or is there something else?

naegelejd replied:
My misconception about compiler optimizations for static vs unnamed namespace. This change has been reverted.

johnstairs commented on cpp/evolution/v2/lib/binary/protocols.cc:
```diff
@@ -0,0 +1,3957 @@
+  case Version::v1: {
+    std::string float_to_string;
+    yardl::binary::ReadString(stream_, float_to_string);
+    value = std::stof(float_to_string);
```
Similarly, this doesn't give an informative error message if `float_to_string` is something like "Hello". It's hard for the user to know what the problem is, or even whether there is a bug in the tool.

johnstairs commented on tooling/internal/cpp/binary/binary.go:
```diff
@@ -84,18 +84,18 @@ func writeHeaderFile(env *dsl.Environment, options packaging.CppCodegenOptions)
 			w.Indented(func() {
 				w.WriteStringln("public:")
-				fmt.Fprintf(w, "%s(std::ostream& stream)\n", writerClassName)
+				fmt.Fprintf(w, "%s(std::ostream& stream, Version version = Version::Latest)\n", writerClassName)
```
The name `Latest` could be misleading. Imagine you have a an existing codebase written against v1. Then v2 comes along and you just want to reference it to genereate the ability to write to it. In that case v1 is more "current" than "latest"

johnstairs commented:
If you run `yardl generate --watch` and change the current model file, any warnings produced briefly appear and then are cleared.

johnstairs commented:
If I run `yardl generate --watch` and change a previous model file, the change is ignored.

johnstairs commented on tooling/pkg/dsl/evolution.go:
```diff
@@ -0,0 +1,1505 @@
+	return fmt.Sprintf("%s%s", baseName, argString)
+}
+
+// Returns the base definition for a NamedType
```
Might want to clarify unless the base is a primitive or generic type parameter

johnstairs commented on tooling/pkg/dsl/evolution.go:
```diff
@@ -0,0 +1,1505 @@
+}
+
+// Determine semantic equivalence between Old and New TypeDefinitions
+func getCrossModelSemanticEquivalence(newTds, oldTds []TypeDefinition) map[string]map[string]bool {
```
Not used?

johnstairs commented:
I think we should warn if a protocol is renamed/removed

johnstairs commented:
If I remove a field from a record, I get a compilation error saying that the variable that we create for writing the value is uninitialized. Also, the message "Removing non-Optional field 'xyz' may result in undefined behavior with previous versions" is not so great. I think we should have defined behavior in all cases.

johnstairs commented on tooling/internal/cpp/binary/binary.go:
```diff
@@ -451,15 +505,258 @@ func writeRwFunctionTemplateDeclaration(t dsl.TypeDefinition, w *formatting.Inde
+					if tc.OldMatches[i] {
+						fmt.Fprintf(w, "%s = std::get<%d>(%s);\n", targetName, i, sourceName)
+					} else {
+						fmt.Fprintf(w, "throw new std::runtime_error(\"Union type incompatible with previous version of model\");\n")
```
is it possible to provide a path or desciption of the union?

johnstairs commented:
Changing a type from 
```
[int, float]
```
to 
```
[int, float]
```

Seems to be allowed but the serialization to the previous version is broken.

Changing from `int` to `int*`, however,  is validation error.

Changing from 
```
[int, bool]
```
to 
```
[int, float]
```
 Results in C++ compilation errors.

johnstairs commented:
When changing from int32 to int64 and encountering an overflow, I think we should raise and error.

johnstairs commented:
Seems like this could be considered compatible:
<img width="549" alt="image" src="https://github.com/microsoft/yardl/assets/339432/54d08d3c-337c-4d7c-9fd3-5ff5281c3590">

johnstairs commented:
Also this could be accepted (with warning of possible overflow):
<img width="531" alt="image" src="https://github.com/microsoft/yardl/assets/339432/49fe06c4-8270-4ac0-8f3e-c52aecef8060">

johnstairs commented:
Same here:
<img width="601" alt="image" src="https://github.com/microsoft/yardl/assets/339432/8c922659-f47e-4478-93fa-5cd16242d459">

johnstairs commented:
If you change the numerical values of an enum, we don't seem to do anything about it in the serializers.

johnstairs commented:
<img width="374" alt="image" src="https://github.com/microsoft/yardl/assets/339432/8ae3ace9-4be3-40ed-8efe-9e64f1dbf268">


Results in compilation error:
```
protocols.cc:70:11: error: 'a' may be used uninitialized [-Werror=maybe-uninitialized]
   70 |     value = std::move(a);
      |     ~~~~~~^~~~~~~~~~~~~~
```

johnstairs commented:
I think this should be accepted:
<img width="391" alt="image" src="https://github.com/microsoft/yardl/assets/339432/11565d6f-f3da-486a-9f49-2840679e1a07">

johnstairs commented:
Swapping the union cases and writing to the previous format seems to write the same binary data:
<img width="388" alt="image" src="https://github.com/microsoft/yardl/assets/339432/c8fb5f3e-1578-4f15-86fb-8b6c4634548d">

johnstairs commented:
I'm wondering why it's ok to add a stream to a protocol but not to remove one.

johnstairs commented:
I think it's getting confused with this:
<img width="340" alt="image" src="https://github.com/microsoft/yardl/assets/339432/0ab75a4c-090a-463e-818f-54c1f4c8bb60">
When writing in the old format, it writes out the same as in the new format, effectively swapping A and B types

naegelejd commented:
I have addressed all of the feedback above except for the following items, which we plan to support in the future:


| Missing support for: | Effort | Explanation |
|-|-|-|
| Scalar <-> Vector/Array conversions | Low | This is a reasonable schema change yardl should support |
| Removing Protocol steps | Medium | This is a reasonable schema change yardl should support |
| Changing generic Record type arguments | High | Requires making type conversion dynamic (e.g. template functions) so that a generic Record serializer can convert fields based on type arguments at runtime |
| Testing runtime conversion errors | Medium | Requires 2+ schema versions, codegen, compilation, and GTest for each scenario under test |

I plan to open an issue for each of these and address them after I finish implementing support for Matlab.

johnstairs commented:
I think we should allow `complexfloat32 <-> complexfloat64`

johnstairs commented on docs/cpp/evolution.md:
```diff
@@ -0,0 +1,244 @@
+
+Yardl supports cross-version compatibility by *statically* generating conversions to and from other versions of your schema (i.e. at yardl codegen time).
+
+In the future, yardl will support *dynamic* compatibility between schema versions, enabling backward- and forward-compatibility between any schema version at runtime.
```
We should also call out that for the moment, we only support this for the binary encoding format using C++.

johnstairs commented:
If I change from a `float32` to an `int64`, the runtime checks are not what I would expect. When writing to the previous version, we check if the integer value falls within the range of a float32, which it always will. Then when reading the older version, we simply do a cast, when that is what might overflow.

naegelejd commented:
> If I change from a `float32` to an `int64`, the runtime checks are not what I would expect. When writing to the previous version, we check if the integer value falls within the range of a float32, which it always will. Then when reading the older version, we simply do a cast, when that is what might overflow.

I will get this right. Too bad this isn't in the standard library: https://www.boost.org/doc/libs/1_84_0/libs/numeric/conversion/doc/html/boost_numericconversion/improved_numeric_cast__.html

johnstairs reviewed (approved):
Ship it!
